python/thunder/rdds/series.py

```diff
@@ -582,18 +582,80 @@ def query(self, inds, var='inds', order='F', isOneBased=True):
 
         return keys, values
 
-    def meanOfRegion(self, keys):
-        """Takes the mean of Series values within a single region specified by the passed keys.
+    def __maskToKeys(self, mask, returnNested=False):
+        """Helper method to validate and convert a binary mask to a set of keys for use in
+        mean of/by region(s).
+
+        If returnNested is true, will return a sequence of sequences of keys, suitable for use
+        in meanByRegions. If returnNested is true and an integer or uint mask is passed, a separate set
+        of keys will be returned for each unique nonzero value in the mask, sorted in numeric order of the
+        mask values.
+
+        If returnNested is false, a single sequence of keys will be returned. The keys will be the indices
+        of all nonzero values of the passed mask.
+
+        Parameters
+        ----------
+        mask: ndarray
+
+        returnNested: boolean, optional, default False
+
+        Returns
+        -------
+        sequence of subscripted indices if returnNested is false,
+        sequence of sequence of subscripted indices if returnNested is true
+        """
+        # argument type checking
+        if not isinstance(mask, ndarray):
+            raise ValueError("Mask should be numpy ndarray, got: '%s'" % str(type(mask)))
+        # check for matching shapes only if we already know our own shape; don't trigger action otherwise
+        if self._dims:
+            if mask.shape != self._dims.count:
+                raise ValueError("Shape mismatch between mask '%s' and series '%s'; shapes must be equal" %
+                                 (str(mask.shape), str(self._dims.count)))
+        from numpy import nonzero, transpose, unique
+
+        def maskToIndices(bmask):
+            return [tuple(idxs) for idxs in transpose(nonzero(bmask))]
+
+        if mask.dtype.kind in ('i', 'u') and returnNested:
+            nestedKeys = []
+            # integer or unsigned int mask
+            for group in unique(mask):
+                if group != 0:
+                    keys = maskToIndices(mask == group)
+                    nestedKeys.append(keys)
+            return nestedKeys
+        else:
+            keys = maskToIndices(mask)
+            if returnNested:
+                return [keys]
+            else:
+                return keys
+
+    def meanOfRegion(self, selection):
+        """Takes the mean of Series values within a single region specified by the passed mask or keys.
+
+        The region for which to take the mean may be specified either by a mask array, or directly by
+        Series keys. If an ndarray is passed as `selection`, then the mean will be taken across all series
+        records corresponding to nonzero elements of the passed mask. (The passed ndarray must have the
+        same shape as series.dims.count, otherwise a ValueError will be thrown.)
```

Comment:
I saw this comment after pushing some changes and was briefly worried that I'd forgotten to address it - but no, for my own future reference: after offline discussion with @freeman-lab, we decided to address this concern by adding additional checks to the series methods, which now by default will throw an error if specified keys are not found in the Series object.